[Merged by Bors] - feat(Topology/ContinuousFunction/CompactlySupported) bundle the class of compactly supported continuous functions

yoh-tanimoto opened the pull request:
give definitions of the class of compactly supported continuous functions and prove basic properties.

Motivation: This class is most natural for the domain of the Riesz-Markov-Kakutani theorem #12290

I copied [Mathlib.Topology.ContinuousFunction.ZeroAtInfty](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/ContinuousFunction/ZeroAtInfty.html#ZeroAtInftyContinuousMap) and am trying to adapt it to compactly supported functions.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+            has_compact_support' := by
+              rw [HasCompactSupport, tsupport]
+              simp only [ContinuousMap.coe_zero, Function.support_zero', closure_empty,
+                isCompact_empty]
+}
```
```suggestion
                isCompact_empty] }
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+theorem coe_mul [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : ⇑(f * g) = f * g :=
+  rfl
+
+@[simp]
```
You don't need the simp attribute on this one, as the previous one will already apply.

yoh-tanimoto replied:
ok I will remove the attribute.

yoh-tanimoto replied:
I removed `@[simp]` from `mul_apply`.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+theorem zero_apply [Zero β] : (0 : C_c(α, β)) x = 0 :=
+  rfl
+
+instance instMul [MulZeroClass β] [ContinuousMul β] : Mul C_c(α, β) :=
```
Don't name your instances in general, unless you have a good reason to do so. Same thing below.

yoh-tanimoto replied:
do you mean that I should remove all the names of the instances? As I took the naming from `ZeroAtInfty`, I'm not sure which one is ok.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+theorem mul_apply [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : (f * g) x = f x * g x :=
+  rfl
+
+-- the product of `(f : C(α, β))` and `(g : C_c(α, β))` is in `C_c(α, β)`
```
```suggestion
-- the product of `f : C(α, β)` and `g : C_c(α, β)` is in `C_c(α, β)`
```
In docstring, don't use parentheses around things which are already in quotes, the quotes make it already clear that it's a self-contained unit. Same thing elsewhere in the file.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+
+-- the product of `(f : C(α, β))` and `(g : C_c(α, β))` is in `C_c(α, β)`
+instance instSMulC [MulZeroClass β] [ContinuousMul β] : SMul C(α, β) C_c(α, β) :=
+  ⟨fun f g => ⟨f * g, HasCompactSupport.mul_left g.2⟩⟩
```
Instead of registering this for `C(α, β)`, could you register this for any `ContinuousMapClass`? Even better, could you register this in the more general setting where you have a type gamma acting continuously on beta, for any `ContinuousMapClass` of functions from alpha to gamma?

yoh-tanimoto replied:
Could you elaborate what I should do here? I suppose I should define `SMul` between `F` and `G` with `ContinuousMapClass F α β` and `CompactlySupportedContinuousMapClass G α β`, but I don't know how to get an object of type `G` from, say, `f : F`, `g : G` and `fun x => (f x) * (g x)`.

sgouezel replied:
I mean something like
```
instance [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] : SMul F C_c(α, β) where
  smul f g :=
    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
```
The space on the right of the action is always `C_c(α, β)`, but the space on the left is any space of continuous functions taking values in a space gamma acting continuously on beta.

Note that for this to work you will need to change the assumptions in the section `DistribMulAction` of the file `Topology.Support`: it should be replaced with the more general
```
section SMulZeroClass

variable [TopologicalSpace α] [Zero M] [SMulZeroClass R M]
variable {f : α → R} {f' : α → M} {x : α}

theorem HasCompactSupport.smul_left (hf : HasCompactSupport f') : HasCompactSupport (f • f') := by
  rw [hasCompactSupport_iff_eventuallyEq] at hf ⊢
  exact hf.mono fun x hx => by simp_rw [Pi.smul_apply', hx, Pi.zero_apply, smul_zero]
#align has_compact_support.smul_left HasCompactSupport.smul_left

end SMulZeroClass
```

yoh-tanimoto replied:
I see, I just changed as you suggested.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+  DFunLike.coe_injective.semigroupWithZero _ coe_zero coe_mul
+
+-- need `[AddMonoid β]` here to apply `HasCompactSupport.add`
+instance instAdd [AddMonoid β] [ContinuousAdd β] : Add C_c(α, β) :=
```
A `AddZeroClass` should be enough, here ane below. For this, can you fix the definition of `HasCompactMulSupport.mul`, requiring only `MulOneClass β`?

yoh-tanimoto replied:
OK, I will edit Topology/Support.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+
+lemma zero_at_infty_of_hasCompactSupport [TopologicalSpace β] [Zero β]
+    (f : C_c(α, β)) :
+    Filter.Tendsto f (Filter.cocompact α) (nhds 0) := by
```
```suggestion
    Filter.Tendsto f (Filter.cocompact α) (𝓝 0) := by
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+variable [UniformSpace β] [UniformSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
+lemma zero_at_infty_of_hasCompactSupport [TopologicalSpace β] [Zero β]
```
This lemma only needs topological spaces, not uniform spaces, so it's a little bit strange to have it in the section `Uniform`. Also, this lemma means that you can (should?) instantiate a `ZeroAtInftyContinuousMapClass` instance here.

yoh-tanimoto replied:
OK, I will register an instance.

sgouezel reviewed (commented):
This looks very good, thanks a lot! Here is a first round of comments on the first half of the file -- don't worry that there are many of them, it's the normal process!

yoh-tanimoto commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+open Filter Metric
+
+/-- `C_c(α, β)` is the type of continuous functions `α → β` with compact support from a topological
+space to a metric space with a zero element.
```
thank you very much for your comments! sorry for my late reply, I will look at them now.
I will change the comment.

yoh-tanimoto commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+    { toFun := f
+      continuous_toFun := f.continuous
+      has_compact_support' := by
+        simp only
```
By browsing the file names, Topology/ContinuousFunction/Compact.lean seems the right file? But this is actually about `MetricSpace`. Is it ok to add the lemma here (for `TopologicalSpace`)?

yoh-tanimoto commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,700 @@
+  map_continuous := map_continuous
+  has_compact_support := by
+    intro f
+    exact IsCompact.of_isClosed_subset isCompact_univ (isClosed_tsupport f)
```
done.

sgouezel commented on Mathlib/Topology/ContinuousFunction/Compact.lean:
```diff
@@ -41,6 +41,17 @@ namespace ContinuousMap
+variable {F γ : Type*} [Zero γ] [TopologicalSpace γ] [FunLike F α γ]
+
+/-- In a compact space `α`, any continuous function has compact support. -/
+theorem isCompact_tsupport_of_CompactSpace (f : F) :
```
The theorem name should match the statement. So it should be in terms of `HasCompactSupport`. Also, by the naming convention words do start with a lowercase letter. So
```suggestion
theorem hasCompactSupport_of_compactSpace (f : F) :
```
But the statement as written is not good, because you can see that it does not really involve `f`, only its coercion to functions. So there is no gain in having a `FunLike` type here, compared to just stating the result for functions.

Also, for now it's in the namespace `ContinuousMap`, while there is no `ContinuousMap` anywhere around in the statement. Better move it outside of this namespace, and probably to another file (`Mathlib.Topology.Support`, maybe?).

yoh-tanimoto replied:
OK, I will move this to `Mathlib.Topology.Support`, changing the name and for `f : α → γ` and pulling it outside the namespace.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,709 @@
+/-!
+# Compactly supported continuous functions
+
+The type of compactly supported continuous functions. When the domain is compact,
+`C(α, β) ≃ C_c(α, β)` via the identity map. When the codomain is a metric space, every continuous
+compactly supported map is a bounded continuous function. When the domain is a locally
+compact space, this type has nice properties.
```
Could you be a little bit more specific in the file-level docstring, with a short introduction saying what the file is about, and then a section introduced with `## Main definitions and results` where you give the name of the main statements, and explain their content? Otherwise, someone just opening the file will need to browse through the whole file to find the name of the statements, and won't have any pointer to help.

yoh-tanimoto replied:
I made the description a bit more specific. I think the main points of this file are the definitions and instances, and I wrote so.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,709 @@
+  type classes (e.g., `TopologicalRing`) are sufficiently generalized.
+-/
+
+
+universe u v w
+
+variable {F : Type*} {α : Type u} {β : Type v} {γ : Type w} [TopologicalSpace α]
```
Most of the time, it's not needed to specify universes in Lean, unless in very specific situations. So I'd just go for
```suggestion
variable {F α β γ : Type*} [TopologicalSpace α]
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,709 @@
+you should parametrize over `(F : Type*) [CompactlySupportedContinuousMapClass F α β] (f : F)`.
+
+When you extend this structure, make sure to extend `CompactlySupportedContinuousMapClass`. -/
+structure CompactlySupportedContinuousMap (α : Type u) (β : Type v) [TopologicalSpace α] [Zero β]
+    [TopologicalSpace β] extends ContinuousMap α β : Type max u v where
```
Same thing, no need to specify the universes.
```suggestion
structure CompactlySupportedContinuousMap (α β : Type*) [TopologicalSpace α] [Zero β]
    [TopologicalSpace β] extends ContinuousMap α β where
```

yoh-tanimoto replied:
I see, I will do this (same here, I followed the way of the old file)

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,709 @@
+    [TopologicalSpace α] [Zero β] [TopologicalSpace β] [FunLike F α β]
+    extends ContinuousMapClass F α β : Prop where
+  /-- Each member of the class has compact support. -/
+  has_compact_support (f : F) : HasCompactSupport f
```
```suggestion
  hasCompactSupport (f : F) : HasCompactSupport f
```
Try to match the symbols. No underscores in `HasCompactSupport`, so no reason why you should introduce one here.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,709 @@
+structure CompactlySupportedContinuousMap (α : Type u) (β : Type v) [TopologicalSpace α] [Zero β]
+    [TopologicalSpace β] extends ContinuousMap α β : Type max u v where
+  /-- The function has compact support . -/
+  has_compact_support' : HasCompactSupport toFun
```
Try to match the symbols. No underscores in HasCompactSupport, so no reason why you should introduce one here.
```suggestion
  hasCompactSupport' : HasCompactSupport toFun
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+
+variable [TopologicalSpace β] (x : α)
+
+instance instZero [Zero β] : Zero C_c(α, β) where
```
no need to name your instances, here and below (unless there is a good reason for it, you should remove all names given to instances in this file).
```suggestion
instance [Zero β] : Zero C_c(α, β) where
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+variable [TopologicalSpace β] [TopologicalSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
+lemma zero_at_infty_of_hasCompactSupport (f : F) :
```
The lemma name is not very good, since `hasCompactSupport` does not appear in the name. Instead, it should be `zero_at_infty_of_compactlySupportedContinuousMapClass`. But it's probably worth it to have a standalone lemma, in another file, saying that a function with compact support tends to zero at infinity -- and to use this lemma here to get a very short proof.

In fact, we already have this lemma, `HasCompactSupport.is_zero_at_infty`, so you should just use it to shorten your proof.

yoh-tanimoto replied:
[HasCompactSupport.is_zero_at_infty](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Compact.html#HasCompactSupport.is_zero_at_infty) assumes` [UniformSpace α]`. I guess this assumption is unnecessary? I wrote a lemma in `Topology/Support`. What should I do with the file `Topology.UniformSpace.Compact`?

sgouezel replied:
Yes, the current assumption if `UniformSpace` is not necessary at all, and the lemma is not in the right file. You should move the lemma to the file `Topology/Support`, but not change its name (and not change its proof, because it already works perfectly fine -- you can see that your version of the proof is a little less efficient).

yoh-tanimoto replied:
I did it, except that I did not copy the `#align` lines. I hope that is fine.

yoh-tanimoto replied:
sorry, I didn't know that it needs to be there after being moved. I restored it.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+
+variable [PseudoMetricSpace β] [Zero β] [FunLike F α β] [CompactlySupportedContinuousMapClass F α β]
+
+protected theorem bounded (f : F) : ∃ C, ∀ x y : α, dist ((f : α → β) x) (f y) ≤ C := by
```
The full name of this theorem is `CompactlySupportedContinuousMap.bounded`. It's a little bit misleading, as it applies not only to `CompactlySupportedContinuousMap`, but in fact to all `CompactlySupportedContinuousMapClass`. So rather call it `CompactlySupportedContinuousMapClass.bounded`, i.e., don't put the results of this section in the `CompactlySupportedContinuousMap` namespace.

yoh-tanimoto replied:
Sorry, I don't know how to do this. Some results are about `F`, but when I move them out of `CompactlySupportedContinuousMap` and `under `CompactlySupportedContinuousMapClass`, some notations such as `f.toBCF` are no longer valid. Could you help?

yoh-tanimoto replied:
actually, in accordance with your comment below, I removed the section `Metric`, so this is no longer a problem.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+
+/-- A continuous function on a compact space has automatically compact support. This is not an
+instance to avoid type class loops. -/
+lemma compactlySupportedContinuousMapClass.ofCompact {G : Type*} [FunLike G α β]
```
```suggestion
lemma _root_.compactlySupportedContinuousMapClass.ofCompact (G : Type*) [FunLike G α β]
```
To avoid the full name `CompactlySupportedContinuousMap.compactlySupportedContinuousMapClass.ofCompact`. Also, make `G` explicit as there is no way to guess it.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+end AlgebraicStructure
+
+section ZeroAtInfty
+
```
You should exit the namespace `CompactlySupportedContinuousMap` for this section, as it has broader application range, and also for the other sections which are about `CompactlySupportedContinuousMapClass`.

yoh-tanimoto replied:
I move the section at the end out of `CompactlySupportedContinuousMap`.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+  intro x y
+  exact hC (Set.mem_range_self x) (Set.mem_range_self y)
+
+
```
```suggestion
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+          ext x
+          simp only [Function.mem_support, ne_eq, star_eq_zero]
+        rw [support_star]
+        exact f.2
+    }
```
```suggestion
        exact f.2 }
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,714 @@
+When `β` is a metric space, then every element of `C_c(α, β)` is bounded, and so there is a natural
+inclusion map `CompactlySupportedContinuousMap.toBCF : C_c(α, β) → (α →ᵇ β)`. Via this map
+`C_c(α, β)` inherits a metric as the pullback of the metric on `α →ᵇ β`. Moreover, this map has
+closed range in `α →ᵇ β` and consequently `C_c(α, β)` is a complete space whenever `β` is complete.
```
I don't think that's true: the closure of functions with compact support is rather functions tending to zero at infinity, so `C_c(α, β)` is not complete. And in fact the topology you're putting here is not the "right" topology on `C_c(α, β)`, so this metric structure should probably not be registered as an instance. Maybe drop this section from the PR, and come back to it later as there is some design to be carefully thought through?

yoh-tanimoto replied:
OK. I made another PR #12629 where I defined a subtype of compactly supported functions. I agree that `C_c` should be given a different topology, maybe the one similar to that for test functions? I removed all the part referring to metric.

yoh-tanimoto commented on Mathlib/Topology/Support.lean:
```diff
@@ -303,13 +303,26 @@ theorem continuous_extend_one [TopologicalSpace β] {U : Set α'} (hU : IsOpen U
 
+section Compact
+
+variable {γ : Type*} [TopologicalSpace α] [CompactSpace α] [One γ] [TopologicalSpace γ]
```
I see, thanks for your suggestion.

yoh-tanimoto commented:
I removed all instance names as suggested, except those that are used later. I guess the mathlib convention has been recently changed? (I copied the most of instance names from `ZeroAtInfty`)

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+    obtain ⟨⟨_, _⟩, _⟩ := f
+    obtain ⟨⟨_, _⟩, _⟩ := g
+    congr
+
```
Can you add here the lemma
```
protected lemma hasCompactSupport (f : C_c(α, β)) : HasCompactSupport f := f.hasCompactSupport'
```
to make sure that `f.hasCompactSupport` works whenever `f : C_c(α, β)`?

yoh-tanimoto replied:
yes, done.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+  map_continuous f := f.continuous_toFun
+  hasCompactSupport f := f.hasCompactSupport'
+
+instance : CoeTC F C_c(α, β) :=
+  ⟨fun f =>
+    { toFun := f
+      continuous_toFun := map_continuous f
+      hasCompactSupport' := hasCompactSupport f }⟩
```
You are in the namespace `CompactlySupportedContinuousMap`, so you shouldn't put anything on a generic `CompactlySupportedContinuousMapClass` here. Instead, once you are done with everything on `CompactlySupportedContinuousMap` (i.e., close to the end of the file, after the line `end CompactlySupportedContinuousMap`), open the namespace `CompactlySupportedContinuousMapClass` and put it there. A good way to see what needs moving is removing `[FunLike F α β] [CompactlySupportedContinuousMapClass F α β]` from the variables line above.

yoh-tanimoto replied:
I move everything referring to `CompactlySupportedContinuousMapClass` under the namespace.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+
+/-! ### Algebraic structure
+
+Whenever `β` has the structure of continuous addtive monoid and a compatible topological structure,
```
```suggestion
Whenever `β` has the structure of continuous additive monoid and a compatible topological structure,
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+  smul f g :=
+    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
+
+@[simp]
+theorem coe_smulc [MulZeroClass β] [ContinuousMul β] (f : C(α, β)) (g : C_c(α, β)) :
+    ⇑(f • g) = f * g :=
+  rfl
+
+@[simp]
+theorem smulc_apply [MulZeroClass β] [ContinuousMul β] (f : C(α, β)) (g : C_c(α, β)) :
+    (f • g) x = f x * g x :=
+  rfl
```
These should rather be stated in the general setting of the instance above.

yoh-tanimoto replied:
What do you mean by "in the general setting"? They are stated just below the definition of the instance `SMul F C_c(α, β)`.

sgouezel replied:
Here, these lemmas are stated for `(f : C(α, β)) (g : C_c(α, β))`. They should be for `(f : F) (g : C_c(α, β))`, where `F` is like in the instance (and the multiplication will be replaced by •, probably).

yoh-tanimoto replied:
I see. `coe_smulc` now becomes `⇑(f • g) = fun x => f x • g x`, as this seems to be the only thing that makes sense. Is this useful?

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+`star : β → β` for which `star (0 : β) = 0`. We don't have quite this weak a typeclass, but
+`StarAddMonoid` is close enough.
+
+The `StarAddMonoid` and `NormedStarGroup` classes on `C_c(α, β)` are inherited from their
```
There is no `NormedStarGroup` any more, right?

yoh-tanimoto replied:
yes. I remove this and also the comment on C⋆-ring.

yoh-tanimoto commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+
+section
+
+variable [T2Space γ] [Zero δ]
```
I tried, there is also an option to assume `[R1Space β]`. I don't know whether this works without any assumption. Which one is better? Or both?

```lean
variable [R1Space β] [Zero δ]

/-- Composition of a continuous function with compact support with a cocompact map from an `[R1Space]`
yields another continuous function with compact support. -/
def comp (f : C_c(γ, δ)) (g : β →co γ) : C_c(β, δ) where
  toContinuousMap := (f : C(γ, δ)).comp g
  hasCompactSupport' := by
    simp only [ContinuousMap.toFun_eq_coe, ContinuousMap.coe_comp, ContinuousMap.coe_coe]
    rw [HasCompactSupport]
    have hsuppfc : (Function.support f)ᶜ ∈ Filter.cocompact γ := by
      rw [Filter.mem_cocompact]
      use tsupport f
      constructor
      · exact f.2
      · simp only [Set.compl_subset_compl, Function.support_subset_iff, ne_eq]
        intro x hx
        push_neg at hx
        exact Set.mem_of_subset_of_mem subset_closure (Function.mem_support.mpr hx)
    apply IsCompact.of_isClosed_subset _ isClosed_closure
    rw [Function.support_comp_eq_preimage]
    obtain ⟨K, hK⟩ := Filter.mem_cocompact.mp (g.cocompact_tendsto' hsuppfc)
    simp only [ContinuousMap.toFun_eq_coe, CocompactMap.coe_toContinuousMap, Set.preimage_compl,
      Set.compl_subset_compl] at hK
    exact IsCompact.of_isClosed_subset (IsCompact.closure hK.1) isClosed_closure (closure_mono hK.2)
```

sgouezel replied:
Does it work for [R1Space gamma]?

yoh-tanimoto replied:
I am not sure. I don't think [IsCompact.isClosed](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#IsCompact.isClosed) is valid for `[R1Space X]`, I think a counterexample is any compact R1Space `X`, choosing a point `x : X`, artificially adding a point `y` , considering the space `X ∪ {y}`and declaring that `x` and `y` are indistinguishable. `X` is still compact (any open cover has a finite subcover), but `X` is clearly not closed. So I don't know how to use `[R1Space γ]`. Any idea?

sgouezel replied:
In fact there is no need for any separation assumption, since the sets we consider (`tsupport _`) are both compact and closed, so their preimages are both compact and closed also, and we never need that the closure of a compact set is compact. I have pushed a proof.

yoh-tanimoto replied:
very nice, thank you!

yoh-tanimoto commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,551 @@
+`NonUnitalRing`, `NormedSpace` and `CstarRing`.
+
+When the domain `α` is compact, `ContinuousMap.liftCompactlySupported` gives the identification
+`C(α, β) ≃ C_c(α, β)`. When the codomain is a metric space, every continuous compactly supported map
+is a bounded continuous function by `bounded`.
```
yes, sorry, I forgot to remove it as well.

yoh-tanimoto commented on Mathlib/Topology/Support.lean:
```diff
@@ -308,15 +308,42 @@ theorem continuous_extend_one [TopologicalSpace β] {U : Set α'} (hU : IsOpen U
+
+/-- In a compact space `α`, any continuous function has compact support. -/
+@[to_additive]
+theorem hasCompactMulSupport_of_compactSpace (f : α → γ) :
```
this gives the error "to_additive: can't transport hasCompactMulSupport.of_compactSpace to itself.". What should I do?

sgouezel replied:
Sorry, it should have been
```suggestion
theorem HasCompactMulSupport.of_compactSpace (f : α → γ) :
```

yoh-tanimoto commented:
I removed the porting note on `HasCompactSupport.is_zero_at_infty`
"-- Porting note: move to src/topology/support.lean once the port is over", as indeed I moved it.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+instance [Zero β] : Zero C_c(α, β) where
+  zero := { toFun := (0 : C(α, β))
+            continuous_toFun := (0 : C(α, β)).2
+            hasCompactSupport' := by
+              rw [HasCompactSupport, tsupport]
+              simp only [ContinuousMap.coe_zero, Function.support_zero', closure_empty,
+                isCompact_empty] }
```
```suggestion
            hasCompactSupport' := by simp [HasCompactSupport, tsupport] }
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+variable [AddGroup β] [TopologicalAddGroup β] (f g : C_c(α, β))
+
+instance : Neg C_c(α, β) where
+  neg f := {  toFun := -f.1
+              continuous_toFun := map_continuous (-f.1)
+              hasCompactSupport' := by
+                rw [HasCompactSupport, tsupport]
+                simp only [ContinuousMap.coe_neg, Function.support_neg']
+                exact f.2 }
```
```suggestion
  neg f := { toFun := -f.1
             continuous_toFun := map_continuous (-f.1)
             hasCompactSupport' := by simpa [HasCompactSupport, tsupport] using f.2 }
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+  rfl
+
+instance : Sub C_c(α, β) where
+  sub f g := {  toFun := f.1 - g.1
+                continuous_toFun := map_continuous (f.1 - g.1)
+                hasCompactSupport' := by
+                  rw [HasCompactSupport, tsupport]
+                  simp only [coe_toContinuousMap]
+                  rw [sub_eq_add_neg]
+                  apply HasCompactSupport.add f.2
+                  rw [HasCompactSupport, tsupport]
+                  simp only [ContinuousMap.coe_neg, Function.support_neg']
+                  exact g.2 }
```
```suggestion
  sub f g := { toFun := f.1 - g.1
               continuous_toFun := map_continuous (f.1 - g.1)
               hasCompactSupport' := by 
                 simpa [sub_eq_add_neg] using HasCompactSupport.add f.2 (-g).2 }
```

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+      continuous_toFun := map_continuous f
+      hasCompactSupport' := hasCompactSupport f }⟩
+
+/-- the product of `f : F` assuming `ContinuousMapClass F α γ` and `ContinuousSMul γ β` and
+`g : C_c(α, β)` is in `C_c(α, β)` -/
+instance [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] : SMul F C_c(α, β) where
+  smul f g :=
+    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
+
+@[simp]
+theorem coe_smulc [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) :
+    ⇑(f • g) = fun x => f x • g x :=
+  rfl
+
+@[simp]
+theorem smulc_apply [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) (x : α):
+    (f • g) x = f x • g x :=
+  rfl
```
These ones should be in the `CompactlySupportedContinuousMap` section, because the main player here is indeed `C_c(α, β)`.

yoh-tanimoto replied:
I move them back.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+
+/-- A continuous function on a compact space has automatically compact support. This is not an
+instance to avoid type class loops. -/
+lemma _root_.compactlySupportedContinuousMapClass.ofCompact (G : Type*) [FunLike G α β]
```
```suggestion
lemma of_compactSpace (G : Type*) [FunLike G α β]
```
You're already in the right namespace.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,554 @@
+
+variable [TopologicalSpace β] [TopologicalSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
```
Could you instead write down the instance that `F` is a `ZeroAtInftyContinuousMapClass`?

yoh-tanimoto replied:
done.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,541 @@
+    ⇑(f • g) = fun x => f x • g x :=
+  rfl
+
+@[simp]
```
```suggestion
```
No need for the simp attribute here, as this can be deduced by simp from the previous one.

sgouezel commented on Mathlib/Topology/Support.lean:
```diff
@@ -327,17 +354,17 @@ theorem HasCompactMulSupport.mul (hf : HasCompactMulSupport f) (hf' : HasCompact
-end DistribMulAction
+end SMulZeroClass
 
 section SMulWithZero
```
In the next section, you can remove `HasCompactSupport.smul_left'`, as it is superseded by the new version of `HasCompactSupport.smul_left`. Well, don't remove it, instead add a deprecated alias to `HasCompactSupport.smul_left`.

yoh-tanimoto replied:
I use `@[deprecated]` and removed the proof, then I leave `#align`. Is this fine?

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,541 @@
+  map_zero' := coe_zero
+  map_add' := coe_add
+
+instance [Zero β] {R : Type*} [Zero R] [SMulWithZero R β] [ContinuousConstSMul R β] :
+    SMul R C_c(α, β) :=
+  ⟨fun r f => ⟨⟨r • ⇑f, Continuous.const_smul f.continuous r⟩, HasCompactSupport.smul_left' f.2⟩⟩
```
```suggestion
instance [Zero β] {R : Type*} [SMulZeroClass R β] [ContinuousConstSMul R β] :
    SMul R C_c(α, β) :=
  ⟨fun r f => ⟨⟨r • ⇑f, Continuous.const_smul f.continuous r⟩, HasCompactSupport.smul_left f.2⟩⟩
```
No need for `SMulWithZero`, you can do with the weaker `[SMulZeroClass R β]`. Can you fix the next lemmas acordingly?

yoh-tanimoto replied:
done.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+
+end
+
+export CompactlySupportedContinuousMapClass (hasCompactSupport)
+
```
```suggestion
export CompactlySupportedContinuousMapClass (hasCompactSupport)

```
I'm not sure it's reasonable to export this in the root namespace.

sgouezel commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+  rfl
+
+theorem smulc_apply [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) (x : α):
```
```suggestion
    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] (f : F) (g : C_c(α, β)) (x : α) :
```

sgouezel reviewed (commented):
This looks very good. Thanks, and thanks for bearing for me!
bors d+

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+theorem eq_of_empty [IsEmpty α] (f g : C_c(α, β)) : f = g :=
+  ext <| IsEmpty.elim ‹_›
+
+/-- A continuous function on a compact space has automatically compact support. -/
```
```suggestion
/-- A continuous function on a compact space automatically has compact support. -/
```

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
+
+@[simp]
+theorem coe_smulc [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
```
Can you explain why this is called `coe_smulc`? I couldn't figure it out. What does the `c` mean? Probably adding a docstring about it would be nice. Sorry if this was addressed in earlier reviews; I didn't read them all.

yoh-tanimoto replied:
"c" is for Continuous. I thought it would be nice to remember that `SMul` is defined for `ContinuousMapClass F α γ`, not just `Funlike F α γ`. But if you don't think it is necessary, I will remove it.

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+  DFunLike.coe_injective.nonUnitalNonAssocRing _ coe_zero coe_add coe_mul coe_neg coe_sub
+    (fun _ _ => rfl) fun _ _ => rfl
+
+instance instNonUnitalRing [NonUnitalRing β] [TopologicalRing β] : NonUnitalRing C_c(α, β) :=
```
Why are this and the following instance named, but none of the others are? Let's be consistent. The new automated instance naming scheme is pretty good, so maybe let's not name them manually.

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+  DFunLike.coe_injective.nonUnitalRing _ coe_zero coe_add coe_mul coe_neg coe_sub (fun _ _ => rfl)
+    fun _ _ => rfl
+
+instance instNonUnitalCommRing [NonUnitalCommRing β] [TopologicalRing β] :
```
ditto

yoh-tanimoto replied:
done.

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+variable [NonUnitalSemiring β] [StarRing β] [TopologicalSpace β] [ContinuousStar β]
+  [TopologicalSemiring β]
+
+instance : StarRing C_c(α, β) :=
```
Can you please add a `TrivialStar` instance under the assumption `[TrivialStar β]`?

yoh-tanimoto replied:
done.

j-loreaux commented on Mathlib/Topology/ContinuousFunction/CompactlySupported.lean:
```diff
@@ -0,0 +1,540 @@
+When the domain `α` is compact, `ContinuousMap.liftCompactlySupported` gives the identification
+`C(α, β) ≃ C_c(α, β)`.
+
+## TODO
+
+* Create more intances of algebraic structures (e.g., `NonUnitalSemiring`) once the necessary
+  type classes (e.g., `TopologicalRing`) are sufficiently generalized.
```
I think you copied this from another file and it's no longer applicable.

yoh-tanimoto replied:
removed.

j-loreaux reviewed (commented):
@yoh-tanimoto Thanks for your hard work on this, and sorry about not reviewing sooner; I've been quite busy. Also, thanks @sgouezel for your reviews!

yoh-tanimoto commented:
@sgouezel thanks to you for reviewing and guiding me! I learnt a lot in this PR.

@j-loreaux thanks for your time! Please let me know about the naming of `coe_smulc` above.

sgouezel commented:
@yoh-tanimoto, when you're done with the comments, you should set the PR back to `awaiting-review`, as otherwise it doesn't show up on the queue and people won't comment on it.

sgouezel commented:
Let's get this in. Thanks!
bors r+
